src/clarity/database/datastore.rs

```diff
@@ -7,35 +9,65 @@ use crate::clarity::types::QualifiedContractIdentifier;
 use crate::clarity::util::hash::Sha512Trunc256Sum;
 use crate::clarity::StacksBlockId;
 use std::collections::HashMap;
+use std::hash::Hash;
 
 #[derive(Clone, Debug)]
 pub struct Datastore {
-    store: HashMap<String, String>,
+    store: HashMap<StacksBlockId, HashMap<String, String>>,
     metadata: HashMap<(String, String), String>,
-    chain_tip: StacksBlockId,
+    open_chain_tip: StacksBlockId,
+    current_chain_tip: StacksBlockId,
     chain_height: u32,
+    height_at_chain_tip: HashMap<StacksBlockId, u32>,
+}
+
+fn height_to_id(height: u32) -> StacksBlockId {
+    let input_bytes = height.to_be_bytes();
+    let mut hasher = Sha512Trunc256::new();
+    hasher.update(input_bytes);
+    let hash = Sha512Trunc256Sum::from_hasher(hasher);
+    StacksBlockId(hash.0)
 }
 
 impl Datastore {
     pub fn new() -> Datastore {
+        let id = height_to_id(0);
+
+        let mut store = HashMap::new();
+        store.insert(id, HashMap::new());
+
+        let mut id_height_map = HashMap::new();
+        id_height_map.insert(id, 0);
+
         Datastore {
-            store: HashMap::new(),
+            store,
             metadata: HashMap::new(),
-            chain_tip: StacksBlockId([0u8; 32]),
+            open_chain_tip: id,
+            current_chain_tip: id,
             chain_height: 0,
+            height_at_chain_tip: id_height_map,
         }
     }
 
     pub fn advance_chain_tip(&mut self, count: u32) -> u32 {
+        let cur_height = self.chain_height;
+
+        for i in 1..=count {
```

Comment:
Creating the blocks lazily could be a bit more tricky (but not unfeasible) - you would have to instantiate this new block when `(at-block ...)` constructs are being evaluated.
I saw a few tests in different protocols, where developers are generating a year of blocks (30-50k blocks would be the order of magnitude we're looking at), for testing things like vesting, APY, etc. It would be interesting to benchmark this family of use cases.

Reply:
Oh wow, I wasn't thinking about cases like that with such large block numbers.

Reply:
@lgalabru I think rather than lazy generation of blocks, we could simply only store blocks where data has changed, and then use a lookup table to map block IDs to the most recent relevant block ID. Does that make sense?

Reply:
I think it makes sense yeah

Reply:
Actually I think what I'm suggesting is the same thing as what Brice suggested...

Reply:
Thanks Reed. The latest update looks good to me.